design/ceph/ceph-telemetry.md

```diff
@@ -0,0 +1,61 @@
+# Adding Rook data to Ceph telemetry
+The solution plan agreed upon with the telemetry team is for the Rook operator to add telemetry to
+the Ceph mon `config-key` database, and Ceph will read each of those items for telemetry retrieval.
+See: https://github.com/rook/rook/pull/10161
+
+## Guidelines from the Ceph telemetry team
+- Users must opt in to telemetry
+- Users must also opt into the new (delta) telemetry items added between Ceph versions
+- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
+  usage of the mon database low.
+
+## Proposed metrics for Rook to collect
+- `rook-version` - Rook version.
+  - Done in https://github.com/rook/rook/pull/10161
+- `rook-kubernetes-version` - Kubernetes version.
+  - This is similar to the already collected os/kernel versions.
+- `rook-num-kubernetes-nodes` - Number of Kubernetes nodes.
+- `rook-ceph-csi-version` - Ceph CSI version.
+- Num CSI plugin pods running - This gives idea of how many k8s nodes are in the k8s cluster that
+  are allowed to use Ceph storage.
+  - It is easiest count plugin pods separately since each DaemonSet is a different label. Plus, it
+    can show if some CSI drivers are ever disabled.
+  - `rook-num-csi-rbd-plugin-pods`
+  - `rook-num-csi-cephfs-plugin-pods`
+  - `rook-num-csi-nfs-plugin-pods`
+- `rook-num-nodes` - Number of nodes on which the cluster has running pods.
+  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
+    count for Rook installs in portable environments
+  - We can get this by counting the number of crash collector pods; however, if any users disable
+    the crash collector, we should report an unknown value, possibly `unknown`.
+- `rook-num-storage-classes` - Number of storage classes related to the current Ceph cluster. We can
+  also break this up by specific kinds of storage classes, but the overall may be useful if new
+  classes are added in the future.
+  - `rook-num-rbd-storage-classes`
+  - `rook-num-cephfs-storage-classes`
+  - `rook-num-nfs-storage-classes`
+  - `rook-num-bucket-storage-classes`
+- `rook-num-storage-class-device-sets` - Number of storage class device sets.
+- `rook-num-portable-storage-class-device-sets` - .
+- `rook-mons-are-pvc` - true/false whether mons are on PVC.
+- `rook-network-provider` - The network provider used for the cluster. e.g., default, host, multus.
+- `rook-external-mode` - true/false if the cluster is in external mode.
+
+**DISCUSS:** Travis suggested: Count of each type of CR: cluster, object, file, object store,
```

Comment:
Maybe we can also think of  osd data,

How much consumption and memory is utilized,
It's already there in the dashboard but this can be good to add with the telemetry too

Reply:
Example telemetry: https://gist.github.com/BlaineEXE/2fafaaeec065803390ebd75b498c1a59

Reply:
memory usage already exists, but would it be useful to record the kubernetes resource constraints for memory and CPU?

Reply:
@richardelling I had considered this, but I opted not to include it. I'm not sure how useful that information really is. We can certainly add it, but I can't think of any good insights that would give us that are more helpful to the Ceph or Rook projects than the size of the usage information already collected by Ceph telemetry. It's also possible the telemetry team would consider this information to be sensitive similarly to how they consider the ceph config values sensitive. 

It's also a hard metric to collect since that can change for OSDs for each device set, and we would need several keys. Even just given the complexity, I think we should only push to add this if we can provide rationale for its use.

Reply:
agree on all points. Speaking for ourselves, we already track and manage it through our existing telemetry service.

Reply:
@richardelling agreed, I'd vote for capturing all of the `resources` from the cluster CR and other applicable CRs.

Reply:
For the original question, as discussed offline we don't necessarily need the counts of all the CRs since most of that can be implied by other metrics (ceph metrics already collected for number of pools, object stores, etc)